redis: add support for redis 6 acls via the `AUTH username password` command

jamesbibby opened the pull request:
Commit Message: Add support for the new Redis 6 ACL auth (AUTH username password) 
Additional Description: Adds an optional username to config and auth commands, and tests
Risk Level: Medium
Testing: Added a number of unit tests, am working on building a docker image locally to test in the wild
Docs Changes: Docs will need to be updated to show the new configuration and support for the AUTH username password command and the downstream_username configuration element
Release Notes:  Adds support for redis 6 acls via the `AUTH username password` command
Fixes #11296

jamesbibby commented:
I still have not tested this past writing unit tests, working on building an image now but I wanted to get some feedback on the approach as it is my first PR to envoy.   I don't know the codebase well so please let me know if things make sense or if I can improve them.

Also I modified both the v2 and v3 apis for redis to include the downstream_auth_username.   Not sure if this was necessary or if just v3 should have been updated.

mattklein123 commented:
@HenryYYang PTAL

jamesbibby commented:
I added ACL to the dictionary so the spelling tests would pass, also cleaned up some error messages/comments.

htuch commented on api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto:
```diff
@@ -24,7 +24,7 @@ option (udpa.annotations.file_status).package_version_status = FROZEN;
 // [#extension: envoy.filters.network.redis_proxy]
 
-// [#next-free-field: 7]
+// [#next-free-field: 8]
```
Since the v2 API is frozen, can you make these changes only to the v3 APIs?

jamesbibby replied:
Yes I can give that a go today.

jamesbibby replied:
@htuch The one question I have is around test.   The tests in these two files are based on a v2 config file format:
github.com/envoyproxy/envoy/blob/master/test/extensions/filters/network/redis_proxy/redis_proxy_integration_test.cc
github.com/envoyproxy/envoy/blob/master/test/extensions/clusters/redis/redis_cluster_integration_test.cc

Do these need new tests that are based on a v3 config format or are these tests only relevant to the v2 api?

htuch commented:
/lgtm api

jamesbibby commented:
@mattklein123 I believe the mac failures were not related to this PR, is there a way to re-trigger the CI or should I push something trivial to kick things off again?

mattklein123 commented:
Let's wait on further pushes until @HenryYYang reviews. @HenryYYang PTAL.

HenryYYang commented on source/extensions/filters/network/redis_proxy/proxy_filter.cc:
```diff
@@ -96,6 +99,25 @@ void ProxyFilter::onAuth(PendingRequest& request, const std::string& password) {
+void ProxyFilter::onAuth(PendingRequest& request, const std::string& username,
+                         const std::string& password) {
+  Common::Redis::RespValuePtr response{new Common::Redis::RespValue()};
+  if (config_->downstream_auth_username_.empty() && config_->downstream_auth_password_.empty()) {
```
I think we need to handle the case where password is config but username is empty.  In this case the "AUTH default <password>" should return OK.

HenryYYang commented on source/extensions/filters/network/redis_proxy/proxy_filter.cc:
```diff
@@ -96,6 +99,25 @@ void ProxyFilter::onAuth(PendingRequest& request, const std::string& password) {
+    connection_allowed_ = true;
+  } else {
+    response->type(Common::Redis::RespType::Error);
+    response->asString() = "ERR invalid ACL auth";
```
Let's use "WRONGPASS invalid username-password pair" like what Redis 6 returns.

jamesbibby commented:
@HenryYYang I fixed some code/format issues to get the CI to pass, added documentation and addressed code review comments.  Should be ready for another look.

@htuch I updated the docs in the API, so I think you may need to give another lgtm for the API. (I updated the expected error when an ACL auth call is incorrect to match the redis 6 response)

htuch commented on api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:
```diff
@@ -234,6 +234,18 @@ message RedisProxy {
+
+  // If a username is provided an ACL style AUTH command will be required with a username and password.
+  // Authenticate Redis client connections locally by forcing downstream clients to issue a `Redis
+  // AUTH command <https://redis.io/commands/auth>`_ with this username and the downstream_auth_password
```
Nit: for fields such as `downstream_auth_password`, either add a `:ref:` or write as `*downstream_auth_password*` if they are proximal (which is the case here).

jamesbibby replied:
Fixed.

htuch commented:
@HenryYYang LMK when you think this is ready to merge, thanks.

HenryYYang commented:
👍

mattklein123 commented:
Sorry needs a master merge.

/wait

jamesbibby commented:
Do you want me to merge master or will you folks take care of that?

mattklein123 commented:
You need to merge master, thank you.

jamesbibby commented:
> You need to merge master, thank you.

Done and pushed

jamesbibby commented:
Fixed a merge issue (comments out of alpha order)

jamesbibby commented:
Thanks for working with me on this.   My first serious c++ contribution in like 2 decades :)
